### PR TITLE
Introduced two web-socket endpoints for workspace master to split JSON-RPC messages

### PR DESCRIPTION
**Pull request by skabashnyuk**
### What does this PR do?
- Add the ability to have several JSON-RPC endpoints in a single J2EE container.
- Each endpoint configured with individual ```RequestProcessorConfigurator.Configuration``` component.
- RequestProcessorConfigurator.Configuration for each endpoint supplied by Provider. To be able to override provider implementation.
- ExecutorService that is constructed for RequestProcessorConfigurator.Configuration is monitored with Prometheus metrics.
- Noticed that even tiny queue (~100) significantly increase throughput and reduce the possibility of rejected messages. 

Rework of https://github.com/eclipse/che/pull/12252

## Test1
Major endpoint 
```
che.core.jsonrpc.processor_max_pool_size=50
che.core.jsonrpc.processor_core_pool_size=3
che.core.jsonrpc.processor_queue_capacity=10000000
```
test 
``` docker run ksmster/json-rpc-loader        -cheurl  ws://che-eclipse-che.192.168.64.31.nip.io/api/websocket        -token=t1         -mnum=100000        -tnum=200```

## Test2
Minor endpoint 
```
che.core.jsonrpc.minor_processor_max_pool_size=100
che.core.jsonrpc.minor_processor_core_pool_size=5
che.core.jsonrpc.minor_processor_queue_capacity=100
```
test 
``` docker run ksmster/json-rpc-loader        -cheurl  ws://che-eclipse-che.192.168.64.31.nip.io/api/websocket-minor        -token=t1         -mnum=100000        -tnum=200```

### What issues does this PR fix or reference?
Related to  https://github.com/eclipse/che/issues/11961

#### Release Notes
n/a

#### Docs PR
n/a

### CODE REVIEW

**Comment by skabashnyuk** [comment]
ci-test

**Comment by skabashnyuk** [comment]
ci-build

**Comment by skabashnyuk** [comment]
ci-test

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 1]
```diff
@@ -567,10 +574,24 @@ che.server.secure_exposer.jwtproxy.auth.loader.path=/_app/loader.html
...
-# in case if pool size would be exceeded message execution will be rejected
-che.core.jsonrpc.processor_max_pool_size=100
 
+## Configuration of major "/websocket" endpoint
+# Maximum size of the JSON RPC processing pool
+# in case if pool size would be exceeded message execution will be rejected
+che.core.jsonrpc.processor_max_pool_size=50
+# Initial json processing pool. Minimum number of threads that used to process major JSON RPC messages.
+che.core.jsonrpc.processor_core_pool_size=3
+# Configuration of queue used to process Json RPC messages. If >0 capacity of LinkedBlockingQueue, <=0 SynchronousQueue
```
Comment (sleshchenko) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: I think 
`If >0 capacity of LinkedBlockingQueue, <=0 SynchronousQueue` part may be improved.
something like 
`If the configured value >0 then it's used as a capacity of LinkedBlockingQueue, otherwise (the value <=0) it is used as a capacity of SynchronousQueue`

File: assembly/assembly-wsmaster-war/src/test/java/org/eclipse/che/WsMasterModuleTest.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.concurrent.ExecutorService;
+import org.eclipse.che.api.deploy.MetricsOverrideBinding;
+import org.eclipse.che.api.deploy.jsonrpc.CheJsonRpcWebSocketConfigurationModule;
+import org.eclipse.che.api.deploy.jsonrpc.CheMajorWebSocketEndpointConfiguration;
+import org.eclipse.che.core.metrics.MetricsModule;
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+public class WsMasterModuleTest {
```
Comment (sleshchenko) [inline assembly/assembly-wsmaster-war/src/test/java/org/eclipse/che/WsMasterModuleTest.java]: Do you think we need this test class?
If yes, then I think `setup` should be part of `shouldInjectMettered` because it not general setup for `WsMasterModule` I think

File: core/che-core-api-core/src/main/java/org/eclipse/che/api/core/jsonrpc/commons/RequestProcessorConfigurator.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import java.util.concurrent.ExecutorService;
+
+/** Request processor configurator */
+public interface RequestProcessorConfigurator {
+
+  /**
+   * Get request processor configuration for specified endpoint.
+   *
+   * @param endpointId endpoint id
+   * @return
```
Comment (sleshchenko) [inline core/che-core-api-core/src/main/java/org/eclipse/che/api/core/jsonrpc/commons/RequestProcessorConfigurator.java]: Please add missing statement or remote `@return` at all

File: core/che-core-api-core/src/main/java/org/eclipse/che/api/core/jsonrpc/commons/RequestProcessorConfigurator.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.core.jsonrpc.commons;
+
+import java.util.concurrent.ExecutorService;
+
+/** Request processor configurator */
```
Comment (sleshchenko) [inline core/che-core-api-core/src/main/java/org/eclipse/che/api/core/jsonrpc/commons/RequestProcessorConfigurator.java]: Looks like dummy javadoc =)

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/bootstrapper/KubernetesBootstrapper.java [hunk 5]
```diff
@@ -221,6 +222,11 @@ private void exec(BiConsumer<String, String> outputConsumer, String... command)
   }
 
   private void exec(String... command) throws InfrastructureException {
+    LOG.debug(
```
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/bootstrapper/KubernetesBootstrapper.java]: Why do we need this log for JSON RPC?

File: wsagent/che-wsagent-core/src/main/java/org/eclipse/che/wsagent/server/CheWebSocketEndpoint.java [hunk 6]
```diff
@@ -28,6 +43,10 @@
 @ServerEndpoint(value = "/wsagent", configurator = GuiceInjectorEndpointConfigurator.class)
 public class CheWebSocketEndpoint extends BasicWebSocketEndpoint {
```
Comment (sleshchenko) [inline wsagent/che-wsagent-core/src/main/java/org/eclipse/che/wsagent/server/CheWebSocketEndpoint.java]: I think it would be better to rename it to make it clear that it is wsagent endpoint

File: wsagent/che-wsagent-core/src/main/java/org/eclipse/che/wsagent/server/WsAgentModule.java [hunk 7]
```diff
@@ -28,6 +32,14 @@
   protected void configure() {
     bind(ApiInfoService.class);
     bind(LivenessProbeService.class);
+    bind(ExecutorService.class)
+        .annotatedWith(Names.named("che.core.jsonrpc.major_executor"))
```
Comment (sleshchenko) [inline wsagent/che-wsagent-core/src/main/java/org/eclipse/che/wsagent/server/WsAgentModule.java]: I think it is better to make `che.core.jsonrpc.major_executor` it as constant of `CheWebSocketEndpoint`  class and also rename it to agent specific. It is not major nor minor, it's just the only agent Enpoint, right?

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ExecutorServiceMetrics.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        .description("The total number of tasks that have been rejected for execution")
+        .baseUnit("tasks")
+        .register(registry);
+  }
+
+  private static class RejectedExecutionHandlerWrapper
+      implements java.util.concurrent.RejectedExecutionHandler {
+
+    private final RejectedExecutionHandler handler;
+    private long counter;
```
Comment (metlos) [inline core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ExecutorServiceMetrics.java]: `volatile`?

File: wsagent/che-wsagent-core/src/main/java/org/eclipse/che/wsagent/server/CheWebSocketEndpoint.java [hunk 9]
```diff
@@ -39,6 +58,67 @@ public CheWebSocketEndpoint(
...
+      implements Provider<ExecutorService> {
+
+    private final ThreadPoolExecutor executor;
+
+    @Inject
+    public CheWebSocketEndpointExecutorServiceProvider(
+        @Named("che.core.jsonrpc.processor_core_pool_size") int corePoolSize,
+        @Named("che.core.jsonrpc.processor_max_pool_size") int maxPoolSize,
+        @Named("che.core.jsonrpc.processor_queue_capacity") int queueCapacity) {
+      ThreadFactory factory =
```
Comment (metlos) [inline wsagent/che-wsagent-core/src/main/java/org/eclipse/che/wsagent/server/CheWebSocketEndpoint.java]: Is there a place common to agent and master where we could put this? Both the `factory` and `executor` are instantiated in the very same way here and in `CheMajorWebSocketEndpointExecutorServiceProvider` that is non-trivial enough for it to be prone to oversight when changed in the future...

File: wsagent/che-wsagent-core/src/main/webapp/WEB-INF/classes/codenvy/che-machine-configuration.properties [hunk 10]
```diff
@@ -57,9 +57,14 @@ project.importer.default_importer_id=git
...
 
-# Maximum size of the json processing pool
+## Configuration of "/websocket" endpoint
+# Maximum size of the JSON RPC processing pool
 # in case if pool size would be exceeded message execution will be rejected
-che.core.jsonrpc.processor_max_pool_size=100
+che.core.jsonrpc.processor_max_pool_size=50
+# Initial JSON RPC processing pool. Minimum number of threads that used to process JSON RPC messages.
+che.core.jsonrpc.processor_core_pool_size=3
+# Configuration of queue used to process Json RPC messages. If >0 capacity of LinkedBlockingQueue, <=0 SynchronousQueue
```
Comment (metlos) [inline wsagent/che-wsagent-core/src/main/webapp/WEB-INF/classes/codenvy/che-machine-configuration.properties]: See https://github.com/eclipse/che/pull/12673/files#r258856382

File: assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/MetricsOverrideBinding.java [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+import javax.inject.Named;
+import javax.inject.Singleton;
+import org.eclipse.che.api.deploy.jsonrpc.CheMajorWebSocketEndpointConfiguration;
+import org.eclipse.che.api.deploy.jsonrpc.CheMajorWebSocketEndpointExecutorServiceProvider;
+import org.eclipse.che.api.deploy.jsonrpc.CheMinorWebSocketEndpointConfiguration;
+import org.eclipse.che.api.deploy.jsonrpc.CheMinorWebSocketEndpointExecutorServiceProvider;
+import org.eclipse.che.core.metrics.ExecutorServiceMetrics;
+
+/**
+ * {@link com.google.inject.Module} that provide alternative implementation of {@link
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/MetricsOverrideBinding.java]: > Generally, I think it is a usual practice to use FQNs in the links that are not imported in the source file otherwise.

I'm OK with that but here these classes are imported. 
Or do you mean that it's always better to use FQN?

File: assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/jsonrpc/CheMinorWebSocketEndpointExecutorServiceProvider.java [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    executor =
+        new ThreadPoolExecutor(
+            corePoolSize,
+            maxPoolSize,
+            60L,
+            SECONDS,
+            queueCapacity > 0 ? new LinkedBlockingQueue<>(queueCapacity) : new SynchronousQueue<>(),
+            factory);
+    executor.setRejectedExecutionHandler(
+        (r, __) -> LOG.error("Message {} rejected for execution", r));
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/jsonrpc/CheMinorWebSocketEndpointExecutorServiceProvider.java]: I like `Executor rejected to handle the message {}`
Because then always will see an actual error message only then request which size is not known.

File: assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/MetricsOverrideBinding.java [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+ * org.eclipse.che.api.deploy.jsonrpc.CheMajorWebSocketEndpointExecutorServiceProvider} and {@link
+ * org.eclipse.che.api.deploy.jsonrpc.CheMinorWebSocketEndpointExecutorServiceProvider}. {@link
+ * java.util.concurrent.ExecutorService} that is returned by this providers will publish metrics to
+ * {@link io.micrometer.prometheus.PrometheusMeterRegistry}
+ */
+public class MetricsOverrideBinding implements Module {
```
[the remark this answers is not in the document]
Comment (skabashnyuk) [reply assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/MetricsOverrideBinding.java:32]: can you rephrase?

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 14]
```diff
@@ -567,10 +574,24 @@ che.server.secure_exposer.jwtproxy.auth.loader.path=/_app/loader.html
 che.server.secure_exposer.jwtproxy.image=eclipse/che-jwtproxy:latest
 che.server.secure_exposer.jwtproxy.memory_limit=128mb
 
-# Maximum size of the json processing pool
-# in case if pool size would be exceeded message execution will be rejected
-che.core.jsonrpc.processor_max_pool_size=100
 
+## Configuration of major "/websocket" endpoint
+# Maximum size of the JSON RPC processing pool
+# in case if pool size would be exceeded message execution will be rejected
```
[the remark this answers is not in the document]
Comment (skabashnyuk) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:580]: can you rephrase?

File: assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/MetricsOverrideBinding.java [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+        @Named("che.core.jsonrpc.minor_processor_queue_capacity") int queueCapacity,
+        PrometheusMeterRegistry meterRegistry) {
+      super(corePoolSize, maxPoolSize, queueCapacity);
+      this.meterRegistry = meterRegistry;
+    }
+
+    @Override
+    public ExecutorService get() {
+      if (executorService == null) {
+        synchronized (MeteredCheMinorWebSocketEndpointExecutorServiceProvider.class) {
```
Comment (metlos) [inline assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/MetricsOverrideBinding.java]: Not that it is likely, but generally speaking synchronizing on objects accessible outside of the class is dangerous because it can cause deadlocks in unsuspecting code outside of control of this class.

Consider:
```java
synchronized(MeteredCheMinorWebSocketEndpointExecutorServiceProvider.class) {
  Thread t = new Thread(() -> new MeteredCheMinorWebSocketEndpointExecutorServiceProvider().get());
  t.start();
  t.join();
}
```

Comment (skabashnyuk) [reply assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/MetricsOverrideBinding.java]: I would rather make ```public synchronized ExecutorService get``` Otherwise it might be a small overengineering of this problem.
@sleshchenko @metlos agree?

File: assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/MetricsOverrideBinding.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (skabashnyuk) [reply assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/MetricsOverrideBinding.java:32]: class description has been updated

File: assembly/assembly-wsmaster-war/src/test/resources/logback-test.xml [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+    available under the terms of the Eclipse Public License 2.0
+    which is available at https://www.eclipse.org/legal/epl-2.0/
+
+    SPDX-License-Identifier: EPL-2.0
+
+    Contributors:
+      Red Hat, Inc. - initial API and implementation
+
+-->
+<configuration>
```
[the remark this answers is not in the document]
Comment (skabashnyuk) [reply assembly/assembly-wsmaster-war/src/test/resources/logback-test.xml:15]: Basically to suppress this output.
```
[INFO] --- maven-surefire-plugin:2.22.1:test (default-test) @ assembly-wsmaster-war ---
[INFO]
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.eclipse.che.MetricsOverrideBindingTest
21:06:59.339 [main] DEBUG org.eclipse.che.core.metrics.FileStoresMeterBinder - Add gauge metric for /dev/disk1s1
21:06:59.344 [main] DEBUG org.eclipse.che.core.metrics.FileStoresMeterBinder - Add gauge metric for devfs
21:06:59.344 [main] DEBUG org.eclipse.che.core.metrics.FileStoresMeterBinder - Add gauge metric for /dev/disk1s4
21:06:59.345 [main] DEBUG org.eclipse.che.core.metrics.FileStoresMeterBinder - Add gauge metric for map -hosts
21:06:59.345 [main] DEBUG org.eclipse.che.core.metrics.FileStoresMeterBinder - Add gauge metric for map auto_home
21:06:59.346 [main] DEBUG org.eclipse.che.core.metrics.FileStoresMeterBinder - Add gauge metric for /Users/sj/Library/Application Support/Google/Android File Transfer/Android File Transfer Agent.app
21:06:59.351 [main] DEBUG org.eclipse.che.core.metrics.ExecutorServiceMetrics - Adding rejection monitoring for java.util.concurrent.ThreadPoolExecutor@11392934[Running, pool size = 1, active threads = 0, queued tasks = 0, completed tasks = 0] che.core.jsonrpc.major_executor
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.69 s - in org.eclipse.che.MetricsOverrideBindingTest
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
[INFO]

```
I'm +0 for this. If you like to I can remove logback-test.xml

**Comment by skabashnyuk** [comment]
ci-test

File: assembly/assembly-wsmaster-war/src/test/resources/logback-test.xml (hunk 16 above)
[the remark this answers is not in the document]
Comment (metlos) [reply assembly/assembly-wsmaster-war/src/test/resources/logback-test.xml:15]: Personally, I'm for keeping it. Just wanted to know why it was added because I saw no reason for it in the new tests.

File: core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/execution/ExecutorServiceProvider.java [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+    executor =
+        new ThreadPoolExecutor(
+            corePoolSize,
+            maxPoolSize,
+            60L,
+            SECONDS,
+            queueCapacity > 0 ? new LinkedBlockingQueue<>(queueCapacity) : new SynchronousQueue<>(),
+            factory);
+    executor.setRejectedExecutionHandler(
+        (r, __) -> LOG.error("Executor rejected to handle the message {}", r));
```
Comment (metlos) [inline core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/execution/ExecutorServiceProvider.java]: If this is a generic executor service provider, then I don't think we can be this specific and call the payload a "message". What about `s/message/payload/`?

Comment (skabashnyuk) [reply core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/execution/ExecutorServiceProvider.java]: ```Executor rejected to handle the task {}```?

**Review by vparfonov — APPROVED** [review]

**Comment by skabashnyuk** [comment]
ci-test

File: core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/execution/ExecutorServiceProvider.java (hunk 17 above)
Comment (metlos) [reply core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/execution/ExecutorServiceProvider.java]: :+1:

**Review by metlos — APPROVED** [review]

File: assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/MetricsOverrideBinding.java [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+  @Singleton
+  public static class MeteredCheMinorWebSocketEndpointExecutorServiceProvider
+      extends CheMinorWebSocketEndpointExecutorServiceProvider {
+    private final PrometheusMeterRegistry meterRegistry;
+
+    private ExecutorService executorService;
+
+    @Inject
+    public MeteredCheMinorWebSocketEndpointExecutorServiceProvider(
+        @Named("che.core.jsonrpc.minor_processor_core_pool_size") int corePoolSize,
```
Comment (sleshchenko) [inline assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/MetricsOverrideBinding.java]: Would be nice to have these properties names as constants in related classes, MajorEndpoint and MinorEndpoint.

File: core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/execution/ExecutorServiceProvider.java [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+    executor =
+        new ThreadPoolExecutor(
+            corePoolSize,
+            maxPoolSize,
+            60L,
+            SECONDS,
+            queueCapacity > 0 ? new LinkedBlockingQueue<>(queueCapacity) : new SynchronousQueue<>(),
+            factory);
+    executor.setRejectedExecutionHandler(
+        (r, __) -> LOG.error("Executor rejected to handle the payload {}", r));
```
Comment (sleshchenko) [inline core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/execution/ExecutorServiceProvider.java]: I think that in the case of `MinorEndpoint` it is not really error but warning or even info.

**Review by sleshchenko — APPROVED** [review]
LGTM
Please take a look my inlined comments

**Comment by skabashnyuk** [comment]
ci-test

File: core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/execution/ExecutorServiceProvider.java (hunk 19 above)
Comment (sleshchenko) [reply core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/execution/ExecutorServiceProvider.java]: IMHO But for MajorEndpoint it should be logged in as ERROR

Comment (skabashnyuk) [reply core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/execution/ExecutorServiceProvider.java]: Will handle that in separate pr.

**Comment by skabashnyuk** [comment]
@eclipse/eclipse-che-qa can you take a look?

**Review by garagatyi — APPROVED** [review]
Looks good!
@ibuziuk since this PR changes configuration of thread pools and JSON RPC messages processing we will need to change our configuration of limits in RhChe at the same time as moving to upstream version that will contain this code

**Comment by ibuziuk** [comment]
@garagatyi good catch, could you please create an issue in rh-che and add it to our prio doc ?

**Comment by garagatyi** [comment]
@ibuziuk sure

**Comment by SkorikSergey** [comment]
Selenium tests execution on Eclipse Che Multiuser on OCP (https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1583//Selenium_20tests_20report/) doesn't show any regression against this Pull Request.
